### PR TITLE
Prototype of results manager

### PR DESCRIPTION
**Pull request by collijk**
Prototype for centralizing the results production by introducing an abstraction where components can specify how the results should be produced and then letting the framework handle stuff from there.

There's a high density here because result production needs to be pretty flexible.  My test case was the disability observer PR'ed here: https://github.com/ihmeuw/vivarium_public_health/pull/205
using the latest state of the bep repository stripped down to include a base population, mortality, and diarrhea.  It produced outputs like:
```
{'MEASURE_ylds_due_to_all_causes_AGE_GROUP_1_to_4_CURRENT_YEAR_2020_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_1_to_4_CURRENT_YEAR_2020_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_1_to_4_CURRENT_YEAR_2021_SEX_Female': 1.2482099550846524,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_1_to_4_CURRENT_YEAR_2021_SEX_Male': 1.311003346855611,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_1_to_4_CURRENT_YEAR_2022_SEX_Female': 1.1608162595992735,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_1_to_4_CURRENT_YEAR_2022_SEX_Male': 1.3604034729690109,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_early_neonatal_CURRENT_YEAR_2020_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_early_neonatal_CURRENT_YEAR_2020_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_early_neonatal_CURRENT_YEAR_2021_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_early_neonatal_CURRENT_YEAR_2021_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_early_neonatal_CURRENT_YEAR_2022_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_early_neonatal_CURRENT_YEAR_2022_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_late_neonatal_CURRENT_YEAR_2020_SEX_Female': 0.5859792686640221,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_late_neonatal_CURRENT_YEAR_2020_SEX_Male': 0.5508407988199642,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_late_neonatal_CURRENT_YEAR_2021_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_late_neonatal_CURRENT_YEAR_2021_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_late_neonatal_CURRENT_YEAR_2022_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_late_neonatal_CURRENT_YEAR_2022_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_post_neonatal_CURRENT_YEAR_2020_SEX_Female': 1.8350359712695008,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_post_neonatal_CURRENT_YEAR_2020_SEX_Male': 2.0012946543282735,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_post_neonatal_CURRENT_YEAR_2021_SEX_Female': 2.4191700925587054,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_post_neonatal_CURRENT_YEAR_2021_SEX_Male': 2.513253751947134,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_post_neonatal_CURRENT_YEAR_2022_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_all_causes_AGE_GROUP_post_neonatal_CURRENT_YEAR_2022_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_1_to_4_CURRENT_YEAR_2020_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_1_to_4_CURRENT_YEAR_2020_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_1_to_4_CURRENT_YEAR_2021_SEX_Female': 16.27130834306779,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_1_to_4_CURRENT_YEAR_2021_SEX_Male': 17.08986505722492,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_1_to_4_CURRENT_YEAR_2022_SEX_Female': 15.13206909834767,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_1_to_4_CURRENT_YEAR_2022_SEX_Male': 17.733830986917457,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_early_neonatal_CURRENT_YEAR_2020_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_early_neonatal_CURRENT_YEAR_2020_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_early_neonatal_CURRENT_YEAR_2021_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_early_neonatal_CURRENT_YEAR_2021_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_early_neonatal_CURRENT_YEAR_2022_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_early_neonatal_CURRENT_YEAR_2022_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_late_neonatal_CURRENT_YEAR_2020_SEX_Female': 7.638658323655999,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_late_neonatal_CURRENT_YEAR_2020_SEX_Male': 7.180603270331674,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_late_neonatal_CURRENT_YEAR_2021_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_late_neonatal_CURRENT_YEAR_2021_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_late_neonatal_CURRENT_YEAR_2022_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_late_neonatal_CURRENT_YEAR_2022_SEX_Male': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_post_neonatal_CURRENT_YEAR_2020_SEX_Female': 23.921004625477426,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_post_neonatal_CURRENT_YEAR_2020_SEX_Male': 26.08830531535071,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_post_neonatal_CURRENT_YEAR_2021_SEX_Female': 31.535610135140274,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_post_neonatal_CURRENT_YEAR_2021_SEX_Male': 32.76205783788229,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_post_neonatal_CURRENT_YEAR_2022_SEX_Female': 0.0,
 'MEASURE_ylds_due_to_diarrheal_diseases_AGE_GROUP_post_neonatal_CURRENT_YEAR_2022_SEX_Male': 0.0,
 'total_population': 10000,
 'total_population_tracked': 10000,
 'total_population_untracked': 0}
```

### CODE REVIEW

File: src/vivarium/framework/engine.py [hunk 1]
```diff
@@ -79,6 +80,7 @@ def __init__(self, model_specification: Union[str, Path, ConfigTree] = None,
         self._events = self._plugin_manager.get_plugin('event')
         self._population = self._plugin_manager.get_plugin('population')
         self._resource = self._plugin_manager.get_plugin('resource')
+        self._results = self._plugin_manager.get_plugin('results')
```
Comment (collijk) [inline src/vivarium/framework/engine.py]: Just some standard wiring to the sim context/ builder in this file.

File: src/vivarium/framework/plugins.py [hunk 2]
```diff
@@ -36,7 +36,11 @@
     },
     'resource': {
         'controller': 'vivarium.framework.resource.ResourceManager',
-        'builder_interface': 'vivarium.framework.resource.ResourceInterface'
```
Comment (collijk) [inline src/vivarium/framework/plugins.py]: Same here in the plugin manager.  Just wiring stuff.

**Comment by collijk** [comment]
Test failure here is a trivial regression test about the simulation context to make us mindful about changing the public api.  Will fix if we decide to go this direction.

File: src/vivarium/framework/results.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
+from collections import defaultdict, Counter
+import itertools
+import typing
+from typing import Any, Callable, Dict, List, Tuple
+
+import pandas as pd
+
+if typing.TYPE_CHECKING:
```
Comment (halfhorst) [inline src/vivarium/framework/results.py]: did you get this working / figure out what the issue was when we tried?

File: src/vivarium/framework/results.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+    @property
+    def name(self):
+        return 'results_manager'
+
+    @property
+    def metrics(self):
+        return self._metrics.copy()
+
+    def setup(self, builder: 'Builder'):
```
Comment (halfhorst) [inline src/vivarium/framework/results.py]: this is type checking I've never seen -- does it gain you anything ?

Comment (halfhorst) [reply src/vivarium/framework/results.py]: nevermind, I see. it's how you fix the issue we had

**Comment by halfhorst** [comment]
Is it correct to say that binners and mappers are hooks for altering the state table in ways that are useful for specific producers?

File: src/vivarium/framework/results.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+            The name of the resulting categorical column.
+        bins
+            The bin edges.
+        labels
+            The labels of the resulting bins.
+        cut_kwargs
+            Additional kwargs to provide to :func:`pandas.cut`.
+
+        """
+        # TODO: Some validation
```
Comment (halfhorst) [inline src/vivarium/framework/results.py]: to include picking up duplication of result_col in all of these state table manipulations ?

File: src/vivarium/framework/results.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+class ResultsContext:
+
+    def __init__(self):
+        # Binners split continuous population columns into categorical ones
+        self._binners = []
+        # Mappers apply a mapping function from a population column to a new column
+        self._mappers = []
+        # keys are event names
+        # values are dicts with key (filter, grouper) value (measure, aggregator, additional_keys)
```
Comment (collijk) [inline src/vivarium/framework/results.py]: Filters rely on users for ordering and may contain typos.

Comment (collijk) [reply src/vivarium/framework/results.py]: typing is a list of tuples for the value

File: src/vivarium/framework/results.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+        builder.event.register_listener('time_step__prepare', self.on_time_step_prepare)
+        builder.event.register_listener('time_step', self.on_time_step)
+        builder.event.register_listener('time_step__cleanup', self.on_time_step_cleanup)
+        builder.event.register_listener('collect_metrics', self.on_collect_metrics)
+
+        builder.value.register_value_modifier('metrics', self.get_results)
+
+    def on_time_step_prepare(self, event: 'Event'):
+        self.gather_results('time_step__prepare', event)
```
Comment (collijk) [inline src/vivarium/framework/results.py]: Maybe just use the event as the key for self._producers in ResultsContext

File: src/vivarium/framework/results.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+            return f'{str(field).upper()}_{param}'
+
+        for params, val in data.iteritems():
+            key = '_'.join(
+                [_format('measure', measure)]
+                + [_format(field, measure) for field, param in zip(data.index.names, params)]
+                # Sorts additional_keys by the field name.
+                + [_format(field, measure) for field, param in sorted(additional_keys.items())]
+            )
+            results[key] = val
```
Comment (collijk) [inline src/vivarium/framework/results.py]: more informative variable names?

File: src/vivarium/framework/results.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+        self._binners = []
+        # Mappers apply a mapping function from a population column to a new column
+        self._mappers = []
+        # keys are event names
+        # values are dicts with key (filter, grouper) value (measure, aggregator, additional_keys)
+        self._producers = defaultdict(lambda: defaultdict(list))
+
+        self.population = None
+
+    def add_binner(self, target_col: str, result_col: str, bins: List, labels: List[str], **cut_kwargs):
```
Comment (collijk) [inline src/vivarium/framework/results.py]: add_cut_argument

Comment (collijk) [reply src/vivarium/framework/results.py]: add_bin_argument

**Pull request closed without merging**